grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandlerBuilder.java

```diff
@@ -0,0 +1,221 @@
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(Iterable<?> messagesOrMessageTypes) {
+        requireNonNull(messagesOrMessageTypes, "messagesOrMessageTypes");
+        if (jsonMarshaller != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register custom messagesOrMessageTypes because a custom JSON marshaller has " +
+                    "already been set. Use the custom marshaller to register custom messages or message types."
+            );
+        }
+
+        for (final Object messageOrMessageType : messagesOrMessageTypes) {
+            requireNonNull(messageOrMessageType, "messagesOrMessageTypes contains null.");
+            if (messageOrMessageType instanceof Message) {
+                messages.add((Message) messageOrMessageType);
+            } else if (messageOrMessageType instanceof Class &&
+                       Message.class.isAssignableFrom((Class<?>) messageOrMessageType)) {
+                messageTypes.add((Class<? extends Message>) messageOrMessageType);
+            } else {
+                final String className =
+                        messageOrMessageType instanceof Class ? ((Class<?>) messageOrMessageType).getName()
+                                                              : messageOrMessageType.getClass().getName();
+                throw new IllegalArgumentException(
+                        className + " is neither Message class nor Message subclass.");
+            }
+        }
+
+        return this;
+    }
+
+    /**
+     * Builds and returns an instance of {@link UnframedGrpcErrorHandler}.
+     *
+     * <p>This method constructs a new {@code UnframedGrpcErrorHandler} with the
+     * current configuration of this builder.
+     */
+    public UnframedGrpcErrorHandler build() {
+        if (jsonMarshaller == null) {
+            jsonMarshaller = UnframedGrpcErrorHandlers.ERROR_DETAILS_MARSHALLER;
+            MessageMarshaller.Builder builder = jsonMarshaller.toBuilder();
+
+            for (final Message message : messages) {
+                builder = builder.register(message);
+            }
+
+            for (final Class<? extends Message> messageType : messageTypes) {
+                builder = builder.register(messageType);
+            }
+
+            jsonMarshaller = builder.build();
+        }
+        if (responseTypes == null) {
+            return UnframedGrpcErrorHandlers.of(statusMappingFunction, jsonMarshaller);
+        }
+        if (responseTypes.equals(ImmutableSet.of(UnframedGrpcErrorResponseType.JSON))) {
+            return UnframedGrpcErrorHandlers.ofJson(statusMappingFunction, jsonMarshaller);
+        }
+        if (responseTypes.equals(ImmutableSet.of(UnframedGrpcErrorResponseType.PLAINTEXT))) {
+            return UnframedGrpcErrorHandlers.ofPlaintext(statusMappingFunction);
+        }
+        return UnframedGrpcErrorHandlers.of(statusMappingFunction, jsonMarshaller);
```

Comment:
Could be optimized into something like this:

```suggestion
        if (responseTypes != null) {
            if (responseTypes.contains(UnframedGrpcErrorResponseType.JSON))) {
                if (!responseTypes.contains(UnframedGrpcErrorResponseType.PLAINTEXT)) {
                    return UnframedGrpcErrorHandlers.ofJson(statusMappingFunction, jsonMarshaller);
                } else {
                    // responseType contains both JSON and PLAINTEXT.
                }
            } else if (responseTypes.contains(UnframedGrpcErrorResponseType.PLAINTEXT)) {
                return UnframedGrpcErrorHandlers.ofPlaintext(statusMappingFunction);
            } else {
                // responseType is empty.
            }
        }
        
        // responseType is null or empty, or contains both JSON and PLAINTEXT.
        return UnframedGrpcErrorHandlers.of(statusMappingFunction, jsonMarshaller);
```